text/0000-const-ub.md

```diff
@@ -0,0 +1,112 @@
+However, this can lead to programs with UB being translated into programs without UB, so if constant evaluation runs after such an optimization, it might fail to detect the UB.
+The only permissible optimizations are those that preserve all UB and that preserve the behavior of programs whose UB CTFE does not detect.
+Formally speaking this means they must be correct optimizations for the abstract machine *that CTFE actually implements*, not just for the abstract machine that specifies Rust; and moreover they must preserve the location and kind of UB that is detected by CTFE.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+To be able to either detect UB or continue evaluation in a well-defined way, CTFE must run on unoptimized code.
+This means when compiling a `const fn` in some crate, the unoptimized code needs to be stored.
+So either the code is stored twice (optimized and unoptimized), or optimizations can only happen after all CTFE results have been computed.
+[Experiments in rustc](https://perf.rust-lang.org/compare.html?start=35debd4c111610317346f46d791f32551d449bd8&end=3dbdd3b981f75f965ac04452739653a3d47ff0ed) showed a severe performance impact on CTFE stress-tests, but no impact on real code except for a slowdown of "incr-unchanged" (which are rather fast so small changes lead to large percentages).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternative is to say that UB during CTFE will definitely be detected.
+However, that is expensive and might even be impossible.
+Even Miri does not currently detect all UB, and Miri is already performing many additional checks that would significantly slow down CTFE.
+Furthermore, implementing these checks requires a more precise understanding of UB than we currently have; basically, this would block having any potentially-UB operations at const-time on having a spec for Rust that precisely describes their UB in a checkable way.
+In particular, this would mean we need to decide on an aliasing model before permitting raw pointers in CTFE.
+
+To avoid the need for keeping the unoptimized sources of `const fn` around, we could weaken the requirement for detecting UB and instead say that UB might cause arbitrary evaluation results.
+Under the assumption that unsound code is not subject to the usual stability guarantees, this is an option we can still move to in the future, should it turn out that the proposal made in this RFC is too expensive.
+
+Another extreme alternative would be to say that UB during CTFE may have arbitrary effects in the host compiler, including host-level UB.
+Basically this would mean that CTFE would be allowed to "leave its sandbox".
+This would allow JIT'ing CTFE and running the resulting code unchecked.
+While compiling untrusted code should only be done with care (including additional sandboxing), this seems like an unnecessary extra footgun.
```

Comment:
Another alternative would be to not allow any unsafe operations in constant contexts

Reply:
That is backwards-incompatible -- we already allow some unsafe operations in `const`/`static` bodies.

Reply:
we have in fact allowed this since before miri, we just didn't allow you to use them for anything transmute-like (you were only able to read from a union field if you also wrote to that union field)